core/src/main/java/io/parsingdata/metal/data/DataExpressionSource.java

```diff
@@ -48,28 +51,32 @@ public DataExpressionSource(final ValueExpression dataExpression, final int inde
     @Override
     protected byte[] getData(final BigInteger offset, final BigInteger length) {
         checkNotNegative(offset, "offset");
-        final Value inputValue = getValue();
-        if (checkNotNegative(length, "length").add(offset).compareTo(inputValue.slice.length) > 0) {
+        final byte[] data = getValue();
+        if (checkNotNegative(length, "length").add(offset).compareTo(BigInteger.valueOf(data.length)) > 0) {
             throw new IllegalStateException("Data to read is not available ([offset=" + offset + ";length=" + length + ";source=" + this + ").");
         }
         final byte[] outputData = new byte[length.intValueExact()];
-        System.arraycopy(inputValue.getValue(), offset.intValueExact(), outputData, 0, outputData.length);
+        System.arraycopy(data, offset.intValueExact(), outputData, 0, outputData.length);
         return outputData;
     }
 
     @Override
     protected boolean isAvailable(final BigInteger offset, final BigInteger length) {
-        return checkNotNegative(offset, "offset").add(checkNotNegative(length, "length")).compareTo(getValue().slice.length) <= 0;
+        return checkNotNegative(offset, "offset").add(checkNotNegative(length, "length")).compareTo(BigInteger.valueOf(getValue().length)) <= 0;
     }
 
-    private Value getValue() {
-        final ImmutableList<Optional<Value>> results = dataExpression.eval(parseState, encoding);
-        if (results.size <= index) {
-            throw new IllegalStateException("ValueExpression dataExpression yields " + results.size + " result(s) (expected at least " + (index + 1) + ").");
+    private synchronized byte[] getValue() {
+        if (cache == null) {
+            final ImmutableList<Optional<Value>> results = dataExpression.eval(parseState, encoding);
+            if (results.size <= index) {
+                throw new IllegalStateException("ValueExpression dataExpression yields " + results.size + " result(s) (expected at least " + (index + 1) + ").");
+            }
+            cache = getValueAtIndex(results, index, 0)
+                .computeResult()
+                .map(Value::getValue)
+                .orElseThrow(() -> new IllegalStateException("ValueExpression dataExpression yields empty Value at index " + index + "."));
         }
-        return getValueAtIndex(results, index, 0)
-            .computeResult()
-            .orElseThrow(() -> new IllegalStateException("ValueExpression dataExpression yields empty Value at index " + index + "."));
+        return Arrays.copyOf(cache, cache.length);
```

Comment:
'nitpick' cache.clone()? But I think that's just a personal preference...

Reply:
Good point. I thought I had some reason for doing it like this with regard to shallow copies on object or multidimensional arrays, but that's all true for `Arrays.copyOf()` as well :)